### PR TITLE
etcdctl/ctlv3: document auth,user,role

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/5580.

### CODE REVIEW

File: etcdctl/README.md [hunk 1]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
 finished transforming keys
 ```
 
+### AUTH <enable or disable>
+
+`auth enable` activates the authentication. `auth disable` to deactivate. Once authentication is on, every client request will be checked on its permission based on its user name.
```
Comment (heyitsanthony) [inline etcdctl/README.md]: `auth enable` activates authentication on an etcd cluster and `auth disable` deactivates. When authentication is enabled, etcd checks all requests for appropriate authorization.

File: etcdctl/README.md [hunk 2]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+
+#### Return value
+
+##### Simple reply
+
+- `Authentication Enabled`. Exit code is zero.
+
+- Error string if PUT failed. Exit code is non-zero.
```
Comment (heyitsanthony) [inline etcdctl/README.md]: if AUTH failed

File: etcdctl/README.md [hunk 3]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+
+``` bash
+etcdctl user add root
+etcdctl user grant-role root root
+etcdctl user get root
+etcdctl role add root
+etcdctl role get root
+etcdctl auth enable
+```
+
```
Comment (heyitsanthony) [inline etcdctl/README.md]: one blank line between between code snippet and header

File: etcdctl/README.md [hunk 4]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl role delete myrole
+```
+
+
```
Comment (heyitsanthony) [inline etcdctl/README.md]: one blank line

File: etcdctl/README.md [hunk 5]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl role add myrole
+```
+
+
```
Comment (heyitsanthony) [inline etcdctl/README.md]: one blank line

File: etcdctl/README.md [hunk 6]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+- Detailed role information. Exit code is zero.
+
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl role get myrole
+Role myrole
+KV Read:
```
Comment (heyitsanthony) [inline etcdctl/README.md]: this needs some data

File: etcdctl/README.md [hunk 7]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+KV Write:
+```
+
+
+### ROLE grant-permission <role name> <permission type> <key> [endkey]
```
Comment (heyitsanthony) [inline etcdctl/README.md]: ROLE GRANT-PERMISSION

File: etcdctl/README.md [hunk 8]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl role delete myrole
+```
+
+
+### ROLE get <role name>
```
Comment (heyitsanthony) [inline etcdctl/README.md]: ROLE GET

File: etcdctl/README.md [hunk 9]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl role add myrole
+```
+
+
+### ROLE delete <role name>
```
Comment (heyitsanthony) [inline etcdctl/README.md]: ROLE DELETE

File: etcdctl/README.md [hunk 10]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+##### Simple reply
+
+- `Role <role name> updated`. Exit code is zero.
+
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl role grant-permission myrole readwrite foo
```
Comment (heyitsanthony) [inline etcdctl/README.md]: output?

File: etcdctl/README.md [hunk 11]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+```
+
+
```
Comment (heyitsanthony) [inline etcdctl/README.md]: one blank line

File: etcdctl/README.md [hunk 12]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+### ROLE revoke-permission <role name> <permission type> <key> [endkey]
```
Comment (heyitsanthony) [inline etcdctl/README.md]: ROLE REVOKE-PERMISSION

File: etcdctl/README.md [hunk 13]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+##### Simple reply
+
+- `Permission of key <key> is revoked from role <role name>`. Exit code is zero.
+
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl role revoke-permission myrole foo
```
Comment (heyitsanthony) [inline etcdctl/README.md]: output?

File: etcdctl/README.md [hunk 14]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+```
+
+
```
Comment (heyitsanthony) [inline etcdctl/README.md]: one blank line

File: etcdctl/README.md [hunk 15]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+### USER add <user name>
```
Comment (heyitsanthony) [inline etcdctl/README.md]: USER ADD

File: etcdctl/README.md [hunk 16]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+
+`user add` to create a user.
```
Comment (heyitsanthony) [inline etcdctl/README.md]: s/to create/creates

File: etcdctl/README.md [hunk 17]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+
+``` bash
+etcdctl user add myuser
+Password of myuser: 
+Type password of myuser again for confirmation: 
+User myuser created
+```
+
+
+### USER delete <user name>
```
Comment (heyitsanthony) [inline etcdctl/README.md]: USER DELETE

File: etcdctl/README.md [hunk 18]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+#### Examples
+
+``` bash
+etcdctl role grant-permission myrole readwrite foo
+```
+
+
+### ROLE revoke-permission <role name> <permission type> <key> [endkey]
+
+`role revoke-permission` to revoke a key from a role.
```
Comment (heyitsanthony) [inline etcdctl/README.md]: s/to revoke/revokes

File: etcdctl/README.md [hunk 19]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+etcdctl role get myrole
+Role myrole
+KV Read:
+KV Write:
+```
+
+
+### ROLE grant-permission <role name> <permission type> <key> [endkey]
+
+`role grant-permission` to grant a key to a role.
```
Comment (heyitsanthony) [inline etcdctl/README.md]: s/to grant/grants

File: etcdctl/README.md [hunk 20]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+#### Examples
+
+``` bash
+etcdctl role delete myrole
+```
+
+
+### ROLE get <role name>
+
+`role get` to get detailed information.
```
Comment (heyitsanthony) [inline etcdctl/README.md]: s/to get detailed information./lists detailed role information.

File: etcdctl/README.md [hunk 21]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+#### Examples
+
+``` bash
+etcdctl role add myrole
+```
+
+
+### ROLE delete <role name>
+
+`role delete` to delete a role.
```
Comment (heyitsanthony) [inline etcdctl/README.md]: s/to delete/deletes

File: etcdctl/README.md [hunk 22]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+etcdctl user get root
+etcdctl role add root
+etcdctl role get root
+etcdctl auth enable
+```
+
+
+### ROLE add <role name>
+
+`role add` to create a role.
```
Comment (heyitsanthony) [inline etcdctl/README.md]: s/to create/creates

File: etcdctl/README.md [hunk 23]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+##### Simple reply
+
+- `User <user name> deleted`. Exit code is zero.
+
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl user delete myuser
```
Comment (heyitsanthony) [inline etcdctl/README.md]: output?

File: etcdctl/README.md [hunk 24]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+etcdctl user add myuser
+Password of myuser: 
+Type password of myuser again for confirmation: 
+User myuser created
+```
+
+
+### USER delete <user name>
+
+`user delete` to delete a user.
```
Comment (heyitsanthony) [inline etcdctl/README.md]: s/to delete/deletes

File: etcdctl/README.md [hunk 25]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl user delete myuser
+```
+
+
```
Comment (heyitsanthony) [inline etcdctl/README.md]: one blank line

File: etcdctl/README.md [hunk 26]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+### USER get <user name>
```
Comment (heyitsanthony) [inline etcdctl/README.md]: USER GET

File: etcdctl/README.md [hunk 27]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+
+`user get` to get detailed information.
```
Comment (heyitsanthony) [inline etcdctl/README.md]: s/to get detailed information/lists detailed user information

File: etcdctl/README.md [hunk 28]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+#### Examples
+
+``` bash
+etcdctl user get myuser
+User: myuser
+Roles:
+```
+
+
+### USER passwd <user name>
```
Comment (heyitsanthony) [inline etcdctl/README.md]: USER PASSWD

File: etcdctl/README.md [hunk 29]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+
+`user passwd` to change password of user
```
Comment (heyitsanthony) [inline etcdctl/README.md]: changes a user's password.

File: etcdctl/README.md [hunk 30]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+#### Examples
+
+``` bash
+etcdctl user passwd myuser
+Password of myuser: 
+Type password of myuser again for confirmation: 
+Password updated
+```
+
+
```
Comment (heyitsanthony) [inline etcdctl/README.md]: one line

File: etcdctl/README.md [hunk 31]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+### USER grant-role <user name> <role name>
```
Comment (heyitsanthony) [inline etcdctl/README.md]: GRANT-ROLE

File: etcdctl/README.md [hunk 32]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+
+`user grant-role` to grant a role to a user
```
Comment (heyitsanthony) [inline etcdctl/README.md]: s/to grant/grants

File: etcdctl/README.md [hunk 33]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+##### Simple reply
+
+- `Role <role name> is granted to user <user name>`. Exit code is zero.
+
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl user grant-role userA roleA
```
Comment (heyitsanthony) [inline etcdctl/README.md]: output?

File: etcdctl/README.md [hunk 34]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+```
+
+
```
Comment (heyitsanthony) [inline etcdctl/README.md]: one blank line

File: etcdctl/README.md [hunk 35]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+### USER revoke-role <user name> <role name>
```
Comment (heyitsanthony) [inline etcdctl/README.md]: REVOKE-ROLE

File: etcdctl/README.md [hunk 36]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+
+`user revoke-role` to revoke a role from from a user
```
Comment (heyitsanthony) [inline etcdctl/README.md]: s/to revoke/revokes
s/from from/from

File: etcdctl/README.md [hunk 37]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+##### Simple reply
+
+- `Role <role name> is revoked from user <user name>`. Exit code is zero.
+
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl user revoke-role userA roleA
```
Comment (heyitsanthony) [inline etcdctl/README.md]: output?

File: etcdctl/README.md [hunk 38]
```diff
@@ -801,6 +801,254 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+```
+
+
```
Comment (heyitsanthony) [inline etcdctl/README.md]: one blank line

**Comment by heyitsanthony** [comment]
overall looks OK, mostly formatting issues

**Comment by gyuho** [comment]
@heyitsanthony PTAL. All formatting issue addressed.

Thanks!

**Comment by heyitsanthony** [comment]
@gyuho the <'s and >'s need to be escaped into \\< and \\> so github will display it

**Comment by gyuho** [comment]
@heyitsanthony Right. Just fixed. PTAL.

Thanks.

**Comment by heyitsanthony** [comment]
lgtm, defer to @xiang90 and @mitake

File: etcdctl/README.md [hunk 39]
```diff
@@ -801,6 +801,252 @@ The provided transformer should read until EOF and flush the stdout before exiti
...
+##### Simple reply
+
+- `Role <role name> deleted`. Exit code is zero.
+
+- Error string if failed. Exit code is non-zero.
+
+#### Examples
+
+``` bash
+etcdctl role delete myrole
```
Comment (xiang90) [inline etcdctl/README.md]: probably we should add the username and password in the etcdctl examples. most of the time, we expect users to do this with auth enabled.

Comment (gyuho) [inline etcdctl/README.md]: ok will add

**Comment by gyuho** [comment]
@xiang90 Just added ` --user=root:123` flags to example. PTAL.

Thanks!

**Comment by xiang90** [comment]
LGTM!

**Comment by gyuho** [comment]
thanks all.

**Comment by mitake** [comment]
@gyuho thanks for doing it!

**Comment by gyuho** [comment]
@mitake Np. Thanks for help!

**Pull request merged**